src/Castle.Core/DynamicProxy/Generators/MethodSignatureComparer.cs

```diff
@@ -120,28 +121,39 @@ private bool EqualSignatureTypes(Type x, Type y, MethodInfo xm = null, MethodInf
 					return false;
 				}
 
-				for (var i = 0; i < xArgs.Length; ++i)
+				if (IsCovariantReturnTypes(x, y, xm, ym) == false)
```

Comment:
This additional check causes a covariant return type generic parameters not to be checked. Is that always handled by `IsAssignableFrom`?

Reply:
I'm not sure I understand. But I'll start with the TL;DR: If you can think of a test case where uninstantiated generic parameters end up being used as the return types, I'm glad to add it to the test suite. I cannot think of such a test case, and I suspect that this might never actually occur in practice.

> This additional check causes a covariant return type generic parameters not to be checked.

When talking about generic parameters, shouldn't we be looking at these bits of code?:

https://github.com/castleproject/Core/blob/451c58fc9510c9faecaf5579fdafa8d65a53e978/src/Castle.Core/DynamicProxy/Generators/MethodSignatureComparer.cs#L89-L92

https://github.com/castleproject/Core/blob/451c58fc9510c9faecaf5579fdafa8d65a53e978/src/Castle.Core/DynamicProxy/Generators/MethodSignatureComparer.cs#L98-L104

I haven't explicitly covered these code branches so far because I wasn't sure that they're actually relevant. Since DynamicProxy only proxies fully instantiated generic types, I suspect we don't actually see (uninstantiated) generic type parameters being used as return types. At least I can't currently think of a test case that would trigger this scenario.

I've now moved my `IsAssignableFrom`-based covariant return type check over to `EqualReturnTypes`, so it will now run even for this scenario (if it actually ever ocurrs). However as you seem to suspect, that won't do much good, because...

> Is that always handled by `IsAssignableFrom`?

`IsAssignableFrom` seems to return `false` when given two generic type parameters, even when they are constrained (via `where T : ...`) such that they are guaranteed to be assignment-compatible.

Reply:
Yer, no idea of a test case to include, because as you said it likely couldn't happen. Happy to resolve the conversation and forget I posted it 😉. Thanks for digging into it, you're always so thorough.

Reply:
... which is mostly thanks to the detailed questions you ask in your code reviews. I've learnt a lot from those. :+1:

I'll ignore generic type parameters for now. If someone finds an actual real-world case where they become a problem, I'll be happy to fix it.